### PR TITLE
ACCUMULO-4813 New bulk import process and API

### PR DESCRIPTION
**Pull request by keith-turner**
This is a big change to how bulk import works.   This new API does the following for bulk import.

 * Computes mapping of files to tablets on the client side.  This mapping is stored in HDFS as JSON (so its human readable). Stored mapping in HDFS so it does need to be read in memory on master.
 * The Fate operation streams the mapping from HDFS (does not load into memory) and metadata table and sends oneway thrift load message to tablet servers.  This makes the fate operation very fast.  Using this new API 10K files were loaded in 20 secs on a single node with 100 tablets and the FATE op took 5 secs.

The new API and functionality is complete, however the following still needs to be done. 

 * Make metadata scan split tolerant (so it never skips a tablet)  The way GC scans is split tolerant.
 * Decide how to make shell support this.
 * Possibly allow user to pass their own mapping file.
 * Retry when concurrent merge happens.
 * Document new API
 * Support bulk import to offline tables
 * When removing load flags, only scan portion of metadata table where loads happened.  Currently scans entire table range in metadata table.

After this PR is finished with review, will open issues for these and anything else that comes up.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java [hunk 1]
```diff
@@ -619,6 +620,63 @@ void setLocalityGroups(String tableName, Map<String,Set<Text>> groups)
...
+   * @since 2.0.0
+   */
+  public static interface ImportSourceArguments {
+    /**
+     *
+     * @param directory
+     *          Load files from this directory
+     * @return ImportSourceOptions
+     */
+    ImportSourceOptions from(String directory);
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java]: I think this should return ImportExecutorOptions

File: core/src/main/java/org/apache/accumulo/core/client/impl/BulkSerialize.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+      } else {
+        return new AbstractMap.SimpleEntry<>(bm.getKeyExtent(tableId), bm.getFiles());
+      }
+    }
+
+  }
+
+  /**
+   * Read Json array of Bulk.Mapping objects and return SortedMap of the bulk load mapping
+   */
```
Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/client/impl/BulkSerialize.java]: Need to update this javadoc

File: core/src/main/java/org/apache/accumulo/core/client/impl/thrift/TableOperationExceptionType.java [hunk 3]
```diff
@@ -30,11 +30,12 @@
...
-  OTHER(6),
-  NAMESPACE_EXISTS(7),
-  NAMESPACE_NOTFOUND(8),
-  INVALID_NAME(9);
+  BULK_BAD_LOAD_MAPPING(5),
+  BAD_RANGE(6),
+  OTHER(7),
+  NAMESPACE_EXISTS(8),
+  NAMESPACE_NOTFOUND(9),
+  INVALID_NAME(10);
```
Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/client/impl/thrift/TableOperationExceptionType.java]: Need to put the new type at the END of the list so the existing numbers don't change.

File: core/src/main/java/org/apache/accumulo/core/master/thrift/FateOperation.java [hunk 4]
```diff
@@ -34,13 +34,14 @@
   TABLE_MERGE(6),
   TABLE_DELETE_RANGE(7),
   TABLE_BULK_IMPORT(8),
-  TABLE_COMPACT(9),
-  TABLE_IMPORT(10),
-  TABLE_EXPORT(11),
-  TABLE_CANCEL_COMPACT(12),
-  NAMESPACE_CREATE(13),
-  NAMESPACE_DELETE(14),
-  NAMESPACE_RENAME(15);
```
Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/master/thrift/FateOperation.java]: Need to put the new type at the END of the list so the existing numbers don't change.

File: server/master/src/main/java/org/apache/accumulo/master/tableOps/TableRangeOpWait.java [hunk 5]
```diff
@@ -39,8 +39,8 @@
  * time on an unmodified tablet.
  *
  * The code below uses read-write lock to prevent some operations while a merge is taking place.
- * Normal operations, like bulk imports, will grab the read lock and prevent merges (writes) while
- * they run. Merge operations will lock out some operations while they run.
+ * Normal operations, like bulkDir imports, will grab the read lock and prevent merges (writes)
+ * while they run. Merge operations will lock out some operations while they run.
```
Comment (milleruntime) [inline server/master/src/main/java/org/apache/accumulo/master/tableOps/TableRangeOpWait.java]: Looks like find/replace caught some unintentional javadoc. We didn't modify this file

**Comment by ctubbsii** [comment]
Since this PR was merged, should the ACCUMULO-4813 JIRA issue be closed now?

**Pull request merged**